### PR TITLE
Enable NullAway for `:core`

### LINKED ISSUE #1680: Consider adopting NullAway into the build process
See: https://github.com/uber/NullAway

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

It would be nice if we can fail our build if there is any potential
`null` dereference, so our users have much less chance of getting
`NullPointerException`.

Modifications:

- Updated the build so that NullAway plugin is enabled for `:core`
  - Note that the root `build.gradle` was modified so that we can
    enable NullAway for other projects in the future.
  - Other caveats:
    - NullAway is disabled for MR-JAR sources because it seems to make
      the build fail with unrelated errors (e.g. no symbol)
- Added `assert` sentences and `@Nullable` annotations wherever
  applicable
- Fixed a few minor potential `null` dereferences

Result:

- We're more confident about our `@Nullable` usages in our API.
- Theoretically, there should be no chance of `NullPointerException` at
  least in the `:core`.
  - However, note that some `NullPointerException`s might have become
    `AssertionError`s, since we added a bunch of assertions in this PR,
    so please review carefully :-)
- Partially resolves #1680 and #5184

### CODE REVIEW

**Comment by minwoox** [comment]
https://github.com/line/armeria/actions/runs/10046097418/job/27764866925?pr=5820#step:11:1515
Could you check the failure?

File: core/src/main/java/com/linecorp/armeria/common/annotation/Nullable.java [hunk 1]
```diff
@@ -32,7 +31,6 @@
  * @see NonNullByDefault
  */
 @Documented
-@Inherited
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/annotation/Nullable.java:35]: 👍

**Review by ikhoon — APPROVED** [review]
Very nice! 👍👍

**Review by minwoox — APPROVED** [review]
👍 👍 👍

**Review by jrhee17 — APPROVED** [review]
👍 👍

**Comment by trustin** [comment]
Thanks, everyone :-)

**Pull request merged**